### PR TITLE
PP-891: Added tests for power provisioning on cray

### PR DESCRIPTION
**Pull request by visheshh**
#### Issue-ID

* **[PP-891](https://pbspro.atlassian.net/browse/PP-891)**

#### Solution description
Added tests for Power provisioning on cray , Tests needs to be run on cray machine.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_power_provisioning_cray.py [hunk 1]
```diff
@@ -56,26 +56,40 @@ def setUp(self):
...
                 self.server.manager(MGR_CMD_SET, NODE,
-                                    {"resources_available.eoe": eoe_vals},
-                                    name)
+                                    {"power_provisioning": True}, n['id'])
                 craynid = n['resources_available.PBScraynid']
                 if min_nid == 0 or craynid < min_nid:
                     min_nid = craynid
                 if craynid > max_nid:
                     max_nid = craynid
+        self.npp = len(self.names) / 3
```
[the remark this answers is not in the document]
Comment (visheshh) [reply test/tests/functional/pbs_power_provisioning_cray.py]: Hi @ashwathraop , Added comment

**Comment by visheshh** [comment]
@PBSPro/pbspro-maintainers , Please review the PR

File: test/tests/functional/pbs_power_provisioning_cray.py [hunk 2]
```diff
@@ -231,28 +263,128 @@ def test_cray_eoe_job(self):
...
+
+        a = {'Resource_List.place': 'scatter',
+             'Resource_List.select': '10:eoe=low+10:eoe=high'}
+        j = Job(TEST_USER, attrs=a)
+        j.set_sleep_time(10)
+        try:
+            self.server.submit(j)
+        except PbsSubmitError as e:
+            self.assertTrue(
+                'Invalid provisioning request in chunk' in e.msg[0])
```
Comment (bhroam) [inline test/tests/functional/pbs_power_provisioning_cray.py]: For negative testing, you need to test both sides.  Right now if the job submits successfully, your test will pass.  I don't think you want this.  say jid = None above the try, set jid to the submit, and do an assertFalse on the jid at the end.

File: test/tests/functional/pbs_power_provisioning_cray.py [hunk 3]
```diff
@@ -231,28 +263,128 @@ def test_cray_eoe_job(self):
...
+        and verify that jobs wont run.
+        """
+
+        self.setup_cray_eoe()   # setup hook and eoe
+        self.enable_power()             # enable hooks
+        eoes = set(['low', 'med', 'high'])
+        a = {'power_provisioning': 'False'}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+
+        while len(eoes) > 0:
```
Comment (bhroam) [inline test/tests/functional/pbs_power_provisioning_cray.py]: Like above, your iterating, use a for loop

File: test/tests/functional/pbs_power_provisioning_cray.py [hunk 4]
```diff
@@ -231,28 +263,128 @@ def test_cray_eoe_job(self):
...
+
+        self.setup_cray_eoe()   # setup hook and eoe
+        self.enable_power()             # enable hooks
+        eoes = set(['low', 'med', 'high'])
+        # set power_provisioning to off where eoe is set to false
+        for i in range(0, self.npp):
+            a = {'power_provisioning': 'False'}
+            self.server.manager(MGR_CMD_SET, NODE, a, id=self.names[i])
+
+        while len(eoes) > 1:
```
Comment (bhroam) [inline test/tests/functional/pbs_power_provisioning_cray.py]: Like above, use a for loop

File: test/tests/functional/pbs_power_provisioning_cray.py [hunk 5]
```diff
@@ -231,28 +263,128 @@ def test_cray_eoe_job(self):
...
+            self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+        jid_low = self.submit_job(10,
+                                  {'Resource_List.place': 'scatter',
+                                   'Resource_List.select': '%d:eoe=%s'
+                                   % (self.npp, 'low')})
+        exp_comm = 'Not Running: Insufficient amount of resource: '
+        exp_comm += 'vntype (cray_compute != cray_login)'
+        self.server.expect(JOB, {
+                           'job_state': 'Q',
+                           'comment': exp_comm}, id=jid_low)
```
Comment (bhroam) [inline test/tests/functional/pbs_power_provisioning_cray.py]: expect() does an or match by default.  You need to say op=PTL_AND if you want to do do an and match.  Either that or use two expects

File: test/tests/functional/pbs_power_provisioning_cray.py [hunk 6]
```diff
@@ -231,28 +263,128 @@ def test_cray_eoe_job(self):
...
+    def test_cray_job_preemption(self):
+        """
+        Submit job to a high priority queue and verify
+        that job is preempted by requeueing.
+        """
+        self.setup_cray_eoe()   # setup hook and eoe
+        self.enable_power()             # enable hooks
+
+        self.server.manager(MGR_CMD_CREATE, QUEUE,
+                            {'queue_type': 'e', 'started': 'True',
```
Comment (bhroam) [inline test/tests/functional/pbs_power_provisioning_cray.py]: expand 'e' to execution

File: test/tests/functional/pbs_power_provisioning_cray.py [hunk 7]
```diff
@@ -56,26 +56,43 @@ def setUp(self):
...
+                                    {"power_provisioning": True}, n['id'])
                 craynid = n['resources_available.PBScraynid']
                 if min_nid == 0 or craynid < min_nid:
                     min_nid = craynid
                 if craynid > max_nid:
                     max_nid = craynid
+        self.npp = len(self.names) / 3
+        # Dividing total number of nodes by 3 and setting each part to a
+        # different power profile , which will be used to submit jobs with
+        # chunks matching to the number of nodes set to each profile
```
Comment (bhroam) [inline test/tests/functional/pbs_power_provisioning_cray.py]: Could this comment go on top of the line where you are dividing by 3?

File: test/tests/functional/pbs_power_provisioning_cray.py [hunk 8]
```diff
@@ -221,7 +253,7 @@ def eoe_job(self, num, eoe):
         self.eoe_check(jid, eoe, secs)
         return jid
 
-    @timeout(700)
+    @timeout(900)
```
Comment (bhroam) [inline test/tests/functional/pbs_power_provisioning_cray.py]: Why the extended timeouts?  Can this test really take 15 minutes?

File: test/tests/functional/pbs_power_provisioning_cray.py [hunk 9]
```diff
@@ -231,28 +263,128 @@ def test_cray_eoe_job(self):
...
+        self.setup_cray_eoe()   # setup hook and eoe
+        self.enable_power()             # enable hooks
+        x = self.npp + 1
+        jid = self.submit_job(10,
+                              {'Resource_List.place': 'scatter',
+                               'Resource_List.select': '%d:eoe=%s' % (x,
+                                                                      'high')})
+        self.server.expect(JOB, {
+            'job_state': 'Q',
+            'comment': 'Not Running: No available resources on nodes'},
```
[the remark this answers is not in the document]
Comment (visheshh) [reply test/tests/functional/pbs_power_provisioning_cray.py]: @bhroam , I checked again autopep8 did not throw any error/warning , I think travis fails if it has any issues with pep8 styling. 
Do you want me to put 'job_state' below expect( ?

**Comment by visheshh** [comment]
@bhroam , Thanks for reviewing , I have addressed your comments please review .

File: test/tests/functional/pbs_power_provisioning_cray.py [hunk 10]
```diff
@@ -57,6 +57,8 @@ def setUp(self):
         if pltfom != 'cray':
             self.skipTest("%s: not a cray")
         self.mom.add_config({"logevent": "0xfffffff"})
+        self.setup_cray_eoe()   # setup hook and eoe
+        self.enable_power()             # enable hooks
```
Comment (bhroam) [inline test/tests/functional/pbs_power_provisioning_cray.py]: Is the indentation off on those two comments?

File: test/tests/functional/pbs_power_provisioning_cray.py [hunk 11]
```diff
@@ -226,15 +226,15 @@ def eoe_check(self, jid, eoe, secs):
...
-        self.mom_logcheck(".*;Job;%s;energy usage*", jid)
+        self.mom_logcheck("capmc get_node_energy_counter --nids")
+        self.mom_logcheck(".*;Job;%s;energy usage", jid)
         self.mom_logcheck(".*;Job;%s;Cray: pcap node", jid)
-        self.mom_logcheck("capmc set_power_cap --nids *")
-        self.mom_logcheck(".*;Job;%s;PMI: reset current_eoe.*", jid)
-        self.mom_logcheck(".*;Job;%s;Cray: remove pcap node*", jid)
+        self.mom_logcheck("capmc set_power_cap --nids")
+        self.mom_logcheck(".*;Job;%s;PMI: reset current_eoe", jid)
+        self.mom_logcheck(".*;Job;%s;Cray: remove pcap node", jid)
```
Comment (bhroam) [inline test/tests/functional/pbs_power_provisioning_cray.py]: No need for the preceding '.*' on the above lines.  log_match() will do substring matches.

File: test/tests/functional/pbs_power_provisioning_cray.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bhroam) [reply test/tests/functional/pbs_power_provisioning_cray.py]: If pep8 is not complaining, don't worry about it.  I just thought pep8 required you to line up everything inside the dictionary.

**Comment by visheshh** [comment]
Hi @bhroam , I have addressed your comments , Please review

**Comment by bhroam** [comment]
After @ashwathraop re-reviews the new changes, please squash and rebase for the merge.

**Comment by visheshh** [comment]
squashed and rebased @bhroam Please take this PR ahead

**Pull request merged**